src/expr/udf/src/external.rs

```diff
@@ -41,25 +46,40 @@ pub struct ArrowFlightUdfClient {
 impl ArrowFlightUdfClient {
     /// Connect to a UDF service.
     pub async fn connect(addr: &str) -> Result<Self> {
-        let conn = tonic::transport::Endpoint::new(addr.to_string())?
-            .timeout(Duration::from_secs(5))
-            .connect_timeout(Duration::from_secs(5))
-            .connect()
-            .await?;
-        let client = FlightServiceClient::new(conn);
-        Ok(Self {
-            client,
-            addr: addr.into(),
-        })
+        Self::connect_inner(
+            addr,
+            ResolutionStrategy::Eager {
+                timeout: TokioDuration::from_secs(5),
+            },
+        )
+        .await
     }
 
-    /// Connect to a UDF service lazily (i.e. only when the first request is sent).
+    /// Connect to a UDF service lazily (i.e. only fetch IP list from DNS when the first request is sent).
     pub fn connect_lazy(addr: &str) -> Result<Self> {
-        let conn = tonic::transport::Endpoint::new(addr.to_string())?
+        block_on(async { Self::connect_inner(addr, ResolutionStrategy::Lazy).await })
+    }
+
+    async fn connect_inner(addr: &str, resolution_strategy: ResolutionStrategy) -> Result<Self> {
+        let addr = addr.strip_prefix("http://").ok_or_else(|| {
+            Error::service_error(format!("udf address must starts with http://: {}", addr))
+        })?;
+        let host_addr = HostAddr::from_str(addr)
+            .map_err(|e| Error::service_error(format!("invalid address: {}, err: {}", addr, e)))?;
```

Comment:
Please avoid to format error directly. May refer to the help message from `check-dylint`.

Reply:
Oops. The lints only work after #15158 gets into main. 😂

Reply:
Updated to `e.as_report()`